lib/oli/delivery/paywall.ex

```diff
@@ -198,28 +198,17 @@ defmodule Oli.Delivery.Paywall do
   end
 
   @doc """
-  Given a section (blueprint or enrollable), calculate the cost to use it for
-  a specific institution, taking into account any product-wide and product-specific discounts
-  this instituttion has.
+    Given a section (blueprint) calculate the cost to use it for
+    a specific institution, taking into account any product-wide and product-specific discounts
+    this instituttion has.
 
-  Returns {:ok, %Money{}} or {:error, reason}
+      Returns {:ok, %Money{}} or {:error, reason}
   """
-  def calculate_product_cost(
-        %Section{requires_payment: false},
-        _
-      ),
-      do: {:ok, Money.new(:USD, 0)}
-
-  def calculate_product_cost(
-        %Section{requires_payment: true, amount: amount},
-        nil
-      ),
-      do: {:ok, amount}
-
-  def calculate_product_cost(
-        %Section{requires_payment: true, id: id, amount: amount},
-        %Institution{id: institution_id}
-      ) do
+  @spec section_cost_from_product(%Section{}, %Institution{}) :: {:ok, %Money{}} | {:error, any}
+   def section_cost_from_product(
+    %Section{requires_payment: true, id: id, amount: amount},
+    %Institution{id: institution_id}
+  ) do
```

Comment:
A couple of comments about the logic inside this function:
1. Nit but I think we could simplify the `where` condition for getting the discounts
```elixir
where: d.institution_id == ^institution_id) and (is_nil(d.section_id) or d.section_id == ^id),
...
```
2. The logic to filter the discounts could be either an `if` instead of a `case`, or something like
```elixir
case Enum.filter(discounts, fn d -> !is_nil(d.section_id) end) do
  [] -> discounts
  filtered_discounts -> filtered_discounts
end
```

Reply:
Gotcha, love the `case` on `Enum.filter` to avoid the unnecessary extra traversal